File: packages/portal-sdk-python/portal_sdk/agent.py
"""Публичный API SDK — класс Agent."""
# ruff: noqa: RUF001
from __future__ import annotations

import json
import os
import sys
from collections.abc import Mapping
from datetime import UTC, datetime
from pathlib import Path
from types import MappingProxyType
from typing import IO, Any

from pydantic import BaseModel, ValidationError
from pydantic_core import PydanticSerializationError

from portal_sdk.events import (
    Artifact,
    ErrorEvent,
    FailedEvent,
    ItemDoneEvent,
    LogEvent,
    LogLevel,
    ProgressEvent,
    ResultEvent,
    StartedEvent,
)

_FINISHED_MSG = "Агент уже завершён (result/failed уже отправлен)."


class Agent:
    """Обёртка над контрактом портала.

    Читает $PARAMS_FILE, $INPUT_DIR, $OUTPUT_DIR из env при инициализации.
    Автоматически отправляет событие `started`; все события — NDJSON в stdout.
    """

    def __init__(self, stdout: IO[str] | None = None) -> None:
        self._stdout = stdout if stdout is not None else sys.stdout
        self._finished = False

        try:
            params_file = Path(os.environ["PARAMS_FILE"])
            self._input_dir = Path(os.environ["INPUT_DIR"])
            self._output_dir = Path(os.environ["OUTPUT_DIR"])
        except KeyError as e:
            raise KeyError(
                f"Обязательная env-переменная не установлена: {e}. "
                "Если ты разрабатываешь агента локально — используй `portal-sdk-run-local`."
            ) from None

        if not params_file.is_file():
            raise FileNotFoundError(f"PARAMS_FILE не найден: {params_file}")

        raw = json.loads(params_file.read_text(encoding="utf-8"))
        if not isinstance(raw, dict):
            raise TypeError(
                f"PARAMS_FILE должен содержать JSON-объект, получен {type(raw).__name__}."
            )
        self._params: dict[str, Any] = raw

        # Сразу публикуем started
        self._emit(StartedEvent(ts=datetime.now(UTC).isoformat()))

    # --- Свойства ---

    @property
    def params(self) -> dict[str, Any]:
        return self._params

    def input_dir(self, input_id: str) -> Path:
        path = self._input_dir / input_id
        if not path.exists():
            raise FileNotFoundError(
                f"Input '{input_id}' не найден в {self._input_dir}. "
                f"Проверь, что в manifest.yaml есть files.{input_id}."
            )
        return path

    @property
    def output_dir(self) -> Path:
        return self._output_dir

    @property
    def env(self) -> Mapping[str, str]:
        """Read-only view на переменные окружения процесса.

        Read-only намеренно: студент не должен случайно мутировать env
        агента через SDK — для этого есть `os.environ` напрямую.
        """
        return MappingProxyType(os.environ)

    # --- События прогресса ---

    def progress(self, value: float, label: str | None = None) -> None:
        """Числовой прогресс 0..1 + опциональная подпись."""
        if self._finished:
            raise RuntimeError(_FINISHED_MSG)
        clamped = max(0.0, min(1.0, value))
        self._emit(ProgressEvent(value=clamped, label=label))

    def log(self, level: str, msg: str) -> None:
        """Сообщение в общую ленту задачи. level: debug|info|warn|error."""
        if self._finished:
            raise RuntimeError(_FINISHED_MSG)
        self._emit(LogEvent(level=LogLevel(level), msg=msg))

    def item_done(
        self,
        item_id: str,
        summary: str | None = None,
        data: dict[str, Any] | None = None,
    ) -> None:
        """Завершение одного элемента в серии (например, одной работы из 46)."""
        if self._finished:
            raise RuntimeError(_FINISHED_MSG)
        self._emit(ItemDoneEvent(id=item_id, summary=summary, data=data))

    def error(self, msg: str, item_id: str | None = None, retryable: bool = True) -> None:
        """Нефатальная ошибка по конкретному элементу. Агент продолжает."""
        if self._finished:
            raise RuntimeError(_FINISHED_MSG)
        self._emit(ErrorEvent(id=item_id, msg=msg, retryable=retryable))

    def result(self, artifacts: list[dict[str, str]]) -> None:
        """Финальное событие успеха.

        artifacts: list[dict] — каждый dict содержит ключи 'id' и 'path'
        (path относительно output_dir). SDK проверяет существование файлов.
        """
        if self._finished:
            raise RuntimeError(_FINISHED_MSG)
        if not artifacts:
            raise ValueError(
                "result() вызван с пустым списком артефактов. "
                "Добавь хотя бы один файл или используй failed() если нечего вернуть."
            )

        normalized = [self._normalize_artifact(a) for a in artifacts]

        # Проверка пути и существования файлов
        for art in normalized:
            # Абсолютные пути запрещены — path должен быть относительным к output_dir
            if Path(art.path).is_absolute():
                raise ValueError(
                    f"Путь артефакта '{art.id}' должен быть относительным к output_dir, "
                    f"получен абсолютный: {art.path!r}. "
                    "Пиши путь как имя файла внутри output_dir, например 'report.docx'."
                )
            full = self._output_dir / art.path
            # Защита от path traversal — путь не должен выходить за output_dir
            try:
                full.resolve().relative_to(self._output_dir.resolve())
            except ValueError:
                raise ValueError(
                    f"Путь артефакта '{art.id}' выходит за пределы output_dir: {art.path!r}. "
                    "Укажи путь относительно output_dir без '..'."
                ) from None
            if not full.is_file():
                raise FileNotFoundError(
                    f"Артефакт '{art.id}' не найден: {full}. "
                    "Перед result() надо записать файл в output_dir."
                )

        self._emit(ResultEvent(artifacts=normalized))
        self._finished = True

    def failed(self, msg: str, details: str | None = None) -> None:
        """Финальное событие неуспеха."""
        if self._finished:
            raise RuntimeError(_FINISHED_MSG)
        self._emit(FailedEvent(msg=msg, details=details))
        self._finished = True

    def _normalize_artifact(self, a: dict[str, str]) -> Artifact:
        if not isinstance(a, dict):
            raise TypeError(
                "artifacts должны быть list[dict] вида [{'id': ..., 'path': ...}], "
                f"получено: {a!r}"
            )
        return Artifact(**a)

    # --- Внутреннее ---

    def _emit(self, event: BaseModel) -> None:
        try:
            line = event.model_dump_json(exclude_none=True)
        except (TypeError, ValueError, ValidationError, PydanticSerializationError) as e:
            raise TypeError(
                f"Поле события {type(event).__name__} не JSON-сериализуемо: {e}. "
                "В data/summary можно класть только числа, строки, bool, None, "
                "списки и словари из этих типов. datetime — Pydantic сам сериализует."
            ) from None
        self._stdout.write(line + "\n")
        self._stdout.flush()

File: packages/portal-sdk-ts/src/agent.ts
/**
 * Agent — обёртка над контрактом портала (TypeScript).
 *
 * Читает $PARAMS_FILE, $INPUT_DIR, $OUTPUT_DIR из env при инициализации.
 * Автоматически отправляет событие `started`. События — NDJSON в stdout.
 */
import { readFileSync, existsSync, statSync } from 'node:fs';
import * as path from 'node:path';
import type {
  AgentEvent,
  Artifact,
  LogLevel,
  StartedEvent,
} from './events.js';

const FINISHED_MSG = 'Агент уже завершён (result/failed уже отправлен).';

/** Куда писать NDJSON. По умолчанию process.stdout. Тестам удобно подменять. */
export interface AgentWriter {
  write(chunk: string): boolean;
}

export interface AgentOptions {
  writer?: AgentWriter;
}

export class Agent {
  readonly params: Record<string, unknown>;
  readonly outputDir: string;

  private readonly _inputDir: string;
  private readonly _writer: AgentWriter;
  private _finished = false;

  constructor(opts: AgentOptions = {}) {
    this._writer = opts.writer ?? process.stdout;
    const paramsFile = process.env.PARAMS_FILE;
    const inputDir = process.env.INPUT_DIR;
    const outputDir = process.env.OUTPUT_DIR;

    if (!paramsFile || !inputDir || !outputDir) {
      const missing = [
        !paramsFile && 'PARAMS_FILE',
        !inputDir && 'INPUT_DIR',
        !outputDir && 'OUTPUT_DIR',
      ]
        .filter(Boolean)
        .join(', ');
      throw new Error(
        `Обязательные env-переменные не установлены: ${missing}. ` +
          'Для локального запуска используй CLI portal-sdk-run-local (Python) или эмулируй env.',
      );
    }
    if (!existsSync(paramsFile)) {
      throw new Error(`PARAMS_FILE не найден: ${paramsFile}`);
    }

    this._inputDir = inputDir;
    this.outputDir = outputDir;

    const raw = JSON.parse(readFileSync(paramsFile, 'utf-8')) as unknown;
    if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
      throw new TypeError(
        `PARAMS_FILE должен содержать JSON-объект, получен ${typeof raw}.`,
      );
    }
    this.params = raw as Record<string, unknown>;

    this._emit({
      type: 'started',
      ts: new Date().toISOString(),
    } satisfies StartedEvent);
  }

  /**
   * Путь к папке с пользовательскими файлами для input_id (из manifest.files).
   * Бросает если папка не найдена (объявленный files-input не пришёл).
   */
  inputDir(inputId: string): string {
    const p = path.join(this._inputDir, inputId);
    if (!existsSync(p)) {
      throw new Error(
        `Input '${inputId}' не найден в ${this._inputDir}. ` +
          `Проверь что в manifest.yaml есть files.${inputId}.`,
      );
    }
    return p;
  }

  /** Числовой прогресс 0..1 + опциональная подпись. */
  progress(value: number, label?: string): void {
    this._guard();
    const clamped = Math.max(0, Math.min(1, value));
    this._emit({ type: 'progress', value: clamped, label: label ?? null });
  }

  /** Сообщение в общую ленту задачи. */
  log(level: LogLevel, msg: string): void {
    this._guard();
    this._emit({ type: 'log', level, msg });
  }

  /** Завершение одного элемента в серии (например, одной работы из 46). */
  itemDone(
    id: string,
    summary?: string,
    data?: Record<string, unknown>,
  ): void {
    this._guard();
    this._emit({
      type: 'item_done',
      id,
      summary: summary ?? null,
      data: data ?? null,
    });
  }

  /** Нефатальная ошибка по элементу. Агент продолжает. */
  error(msg: string, itemId?: string, retryable = true): void {
    this._guard();
    this._emit({ type: 'error', id: itemId ?? null, msg, retryable });
  }

  /**
   * Финальное событие успеха. Artifacts — список `{id, path}` где
   * path относителен outputDir; SDK проверит что файлы существуют.
   */
  result(artifacts: Artifact[]): void {
    this._guard();
    if (artifacts.length === 0) {
      throw new Error(
        'result() вызван с пустым списком артефактов. ' +
          'Добавь хотя бы один файл или используй failed() если нечего вернуть.',
      );
    }
    for (const a of artifacts) {
      if (path.isAbsolute(a.path)) {
        throw new Error(
          `Путь артефакта '${a.id}' должен быть относительным к outputDir, ` +
            `получен абсолютный: ${a.path}.`,
        );
      }
      const full = path.resolve(this.outputDir, a.path);
      const root = path.resolve(this.outputDir);
      if (!full.startsWith(root + path.sep) && full !== root) {
        throw new Error(
          `Путь артефакта '${a.id}' выходит за пределы outputDir: ${a.path}.`,
        );
      }
      if (!existsSync(full) || !statSync(full).isFile()) {
        throw new Error(
          `Артефакт '${a.id}' не найден: ${full}. ` +
            'Перед result() надо записать файл в outputDir.',
        );
      }
    }
    this._emit({ type: 'result', artifacts });
    this._finished = true;
  }

  /** Финальное событие неуспеха. */
  failed(msg: string, details?: string): void {
    this._guard();
    this._emit({ type: 'failed', msg, details: details ?? null });
    this._finished = true;
  }

  private _guard(): void {
    if (this._finished) throw new Error(FINISHED_MSG);
  }

  private _emit(event: AgentEvent): void {
    const clean = JSON.stringify(event, (_, v) => (v === null ? null : v));
    this._writer.write(clean + '\n');
  }
}
